packages/astro/src/integrations/index.ts

```diff
@@ -178,6 +179,31 @@ export async function runHookConfigDone({
 								`Integration "${integration.name}" conflicts with "${settings.adapter.name}". You can only configure one deployment integration.`
 							);
 						}
+						if (!adapter.supportedAstroFeatures) {
+							// NOTE: throw an error in Astro 4.0
+							warn(
+								logging,
+								'astro',
+								`The adapter ${adapter.name} doesn't provide a feature map. From Astro 3.0, an adapter 
+		can provide a feature map. Not providing a feature map will cause an error in Astro 4.0.`
```

Comment:
nit: I think the string should be a single line so it can be line-wrapped in the terminal naturally.